### PR TITLE
Make eslint-plugin validate dependencies in `useSelect` and `useSuspenseSelect` hooks

### LINKED ISSUE #49897: eslint-plugin does not validate dependencies in `useSelect` and `useSuspenseSelect` hooks
## What problem does this address?

[`useSelect`](https://github.com/WordPress/gutenberg/blob/8be5532b62056a4a7b0c5999f458934c8120a516/packages/data/src/components/use-select/index.js#L254-L275) and [`useSuspenseSelect`](https://github.com/WordPress/gutenberg/blob/8be5532b62056a4a7b0c5999f458934c8120a516/packages/data/src/components/use-select/index.js#L292-L294) hooks in https://www.npmjs.com/package/@wordpress/data have dependencies array as their second parameters. The dependencies array is eventually passed to `useCallback` in:

https://github.com/WordPress/gutenberg/blob/8be5532b62056a4a7b0c5999f458934c8120a516/packages/data/src/components/use-select/index.js#L183

For normal React hooks like `useCallback` above, we already have [`'react-hooks/exhaustive-deps': 'warn'`](https://github.com/WordPress/gutenberg/blob/8be5532b62056a4a7b0c5999f458934c8120a516/packages/eslint-plugin/configs/react.js#L37) in https://www.npmjs.com/package/@wordpress/eslint-plugin to help us declare all the dependencies. 

However, the eslint-plugin does not validate dependencies for our custom `useSelect` and `useSuspenseSelect` hooks. This causes issues when developers miss a dependency when using those hooks.

To see this issue in action, find a `useSelect` usage in the code base, and add or remove an item in its dependencies array. There is no eslint error or warning about the dependencies. Note that you may need to reload the editor window to re-run eslint.

Here's an example: when we remove the `clientId` in the dependency array in the following `useSelect` usage, there is no warning or error.

https://github.com/WordPress/gutenberg/blob/361f257b6d36ba14e5c6c5772718d118325a72ef/packages/block-library/src/search/edit.js#L83-L93

image

## What is your proposed solution?

Make the eslint-plugin validate all the dependencies in `useSelect` and `useSuspenseSelect` hooks. 

This can be done by modifying the following line:

https://github.com/WordPress/gutenberg/blob/8be5532b62056a4a7b0c5999f458934c8120a516/packages/eslint-plugin/configs/react.js#L37

to:

```
		'react-hooks/exhaustive-deps': [
			'warn',
			{
				additionalHooks: '(useSelect|useSuspenseSelect)',
			},
		],
```

Reference: https://www.npmjs.com/package/eslint-plugin-react-hooks#advanced-configuration

### PR DESCRIPTION
**Pull request by ecgan**
## What?

Closes https://github.com/WordPress/gutenberg/issues/49897. Builds on top of PR https://github.com/WordPress/gutenberg/pull/24914 that was just merged recently in February 2023.

In this PR, we add `useSelect` and `useSuspenseSelect` as additional hooks in `'react-hooks/exhaustive-deps'` eslint rule in eslint-plugin. 

## Why?

This helps us to declare all the dependencies when we use `useSelect` and `useSuspenseSelect` hooks. When we miss a dependency, there will be an eslint warning. It allows us to make use of the editor's quick fix command to quickly fix the dependency issue.

## Testing Instructions

1. Find a `useSelect` hook with dependency. For example: 

https://github.com/WordPress/gutenberg/blob/361f257b6d36ba14e5c6c5772718d118325a72ef/packages/block-library/src/search/edit.js#L83-L93

2. Add or remove an item in the dependency.
3. There should be an eslint warning. See example screenshot below with "missing dependency" warning when we remove the `clientId` in the dependency array:

image

5. Use the editor's quick fix command to automatically fix the warning.

### CODE REVIEW

**Comment by ecgan** [comment]
The Static Analysis GitHub Actions failed because it has an ESLint error with a `useSelect` usage with missing dependency: https://github.com/WordPress/gutenberg/actions/runs/4734770383/jobs/8404141142?pr=49900#step:5:606. The `useSelect` code usage was introduced in PR https://github.com/WordPress/gutenberg/pull/23922.

It is an ESLint error, not a warning, because it is set to `'react-hooks/exhaustive-deps': 'error'` for the components package in the following file:

https://github.com/WordPress/gutenberg/blob/7b8395e9f6ff11f4ce39053d42cd6adde0fb1839/.eslintrc.js#L353-L364

**Comment by Mamaduka** [comment]
> The Static Analysis GitHub Actions failed because it has an ESLint error with a useSelect usage with missing dependency: https://github.com/WordPress/gutenberg/actions/runs/4734770383/jobs/8404141142?pr=49900#step:5:606. The useSelect code usage was introduced in PR https://github.com/WordPress/gutenberg/pull/23922.

That's [a very unusua](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/link-picker/link-picker-results.native.js#L37-L77)l use of the `useSelect` hook. I think the hook should just return `__experimentalFetchLinkSuggestions`, and then `fetchLinkSuggestions` should be implemented outside of the hook.

cc @WordPress/native-mobile

File: packages/components/src/mobile/link-picker/link-picker-results.native.js
Comment (ecgan) [inline packages/components/src/mobile/link-picker/link-picker-results.native.js:79]: As per https://github.com/WordPress/gutenberg/pull/23922#discussion_r1170634879, I have temporarily disabled eslint rule for this line to suppress eslint error for now.

**Review by ecgan — COMMENTED** [review]
I made an additional code change to fix the eslint error for now. See below.

File: packages/components/src/mobile/link-picker/link-picker-results.native.js
Comment (dcalhoun) [reply packages/components/src/mobile/link-picker/link-picker-results.native.js:79]: Disabling the linter error on mobile-specific violations has been done in the past. I think this makes sense in this case to unblock this PR. Thanks for [the ping](https://github.com/WordPress/gutenberg/pull/49900#issuecomment-1514155916)!

**Comment by ecgan** [comment]
Hi fellow reviewers @dcalhoun @gziolo @ntwb @nerrad @ajitbohra , may I have your review and approval on this PR please? 🙏

**Review by nerrad — APPROVED** [review]
Looks good to me. I don't see any new _errors_ popping up when running linting JS locally and agree this should remain a warning for now in most contexts (although in general I think we should move towards this being an error for all hooks with dependency args at some point).

**Review by dcalhoun — APPROVED** [review]
LGTM. I did a quick test of the mobile Demo editor using an iPhone 14 Pro simulator, I did not encounter any issues. 🚀 

Thanks again for involving the mobile team! 🙇🏻

**Comment by ecgan** [comment]
@nerrad @dcalhoun , thanks so much for the approval. Apparently I am not authorized to merge the PR (see screenshot below). Are you able to grant me the permission? I made a [request in WordPress Slack](https://wordpress.slack.com/archives/C02QB2JS7/p1682006164063609) to join the Gutenberg github team last week. Or perhaps you could help to merge this?

image

**Comment by nerrad** [comment]
@ecgan I just realized, one thing that is missing here is an update to the eslint plugin changelog. Will you submit a followup to take care of that please?

**Comment by ecgan** [comment]
@nerrad , sure, I'll take care of that. Thank you for checking on it. 🙂 🙏

**Comment by rowasc** [comment]
@nerrad @ecgan would it make sense to have a similar setup as the one used in the [flaky test reporter](https://github.com/WordPress/gutenberg/tree/trunk/packages/report-flaky-tests) to [report](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BType%5D+Flaky+Test%22) open `eslint` warnings so that the transition time between introducing a "warn" rule and being able to enable "error" is potentially shorter (or at least easy to track)? 
 
I'm not sure if it's worth it, but I can take a stab at it if this problem is something that comes up often enough and if tracking it would indeed be helpful (I'm not sure if it'd help or just be annoying :) )

**Comment by nerrad** [comment]
🤔  I think it depends on how many issues would get created - I see the flaky test-reported issues _do_ get handled but there are also quite a few older ones still lingering around. Probably a good first step would be to see how many warnings exist in the codebase and if small enough, try to fix them in one (or a set of) PR(s).

**Comment by rowasc** [comment]
yea @nerrad - I got a little overwhelmed by warnings (106 by react-hooks/exhaustive-deps) so I figured it may be a big enough issue that it makes sense for it to be tracked, but I'm not familiar enough yet with how things are handled in Gutenberg to really know if that's "normal" or common enough. The downside of such a tool may be accidentally incentivizing "leaving warnings for later" as a behavior instead of incentivizing fixing them promptly. 

edit: of course, right now we're more in the "leave it for later" situation since warnings are non-blocking for a merge.

**Pull request merged**
